### PR TITLE
Convert publish.gradle to Kotlin

### PR DESCRIPTION
**Pull request by hansd**
This is an interesting one. I'm looking forward to the code review. I intentionally left the structure more or less as it was in the script. I think there is a lot we can do to improve the structure. There are also a couple of areas where I'm not sure what is considered idiomatic Kotlin DSL. I also got rid of two if statements in the root build.gradle. I think our goal should be to have not a single if statement in any of our build.gradle(.kts) scripts.

### CODE REVIEW

File: buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    publishCompile.allDependencies.withType<ProjectDependency>().forEach {
+                        publishRuntime("org.gradle:${it.dependencyProject.base.archivesBaseName}:${version}")
+                    }
+                    publishCompile.allDependencies.withType<ExternalDependency>().forEach {
+                        publishRuntime(it)
+                    }
+                }
+
+                val install by tasks.getting(Upload::class)
+                val mavenInstaller by install.repositories
```
Comment (bamboo) [inline buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt]: :x: `mavenInstaller` is a property of `MavenRepositoryHandlerConvention`.

Something like the following would do:
```kotlin
install.repositories {
  withConvention(MavenRepositoryHandlerConvention::class) {
    mavenInstaller {
      pom.scopeMappings.mappings.clear()
      ...
    }
  }
}
```

File: buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import org.gradle.api.internal.artifacts.publish.DefaultPublishArtifact
+import org.gradle.api.tasks.Upload
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.kotlin.dsl.*
+import java.io.File
+import java.util.*
+
+open class PublishPublicLibrariesPlugin : Plugin<Project> {
+
+    override
```
Comment (bamboo) [inline buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt]: :x: In the `kotlin-dsl` style, the semantic modifiers `override` / `open` / `operator` and `infix` go in the same line as `fun`.

File: buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt (hunk 1 above)
Comment (hansd) [reply buildSrc/src/main/kotlin/org/gradle/plugins/publish/PublishPublicLibrariesPlugin.kt]: Very nice. I have applied this.

**Comment by hansd** [comment]
To be continued in https://github.com/gradle/gradle/pull/4556

**Pull request closed without merging**